File: day1/objc/day1-part2.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *filePath = @"../input.txt";
        NSString *fileContents = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
        
        NSArray *pairs = [fileContents componentsSeparatedByString:@"\n"];
        
        NSMutableArray *list1 = [NSMutableArray array];
        NSMutableArray *list2 = [NSMutableArray array];
        
        for (NSString *pair in pairs) {
            NSArray *components = [pair componentsSeparatedByString:@"   "];
            
            [list1 addObject:@([components[0] integerValue])];
            [list2 addObject:@([components[1] integerValue])];
        }
        
        NSComparisonResult (^ascendingComparison)(NSNumber *num1, NSNumber *num2) = ^NSComparisonResult(NSNumber *num1, NSNumber *num2) {
            if ([num1 integerValue] < [num2 integerValue]) {
                return NSOrderedAscending;
            } else if ([num1 integerValue] > [num2 integerValue]) {
                return NSOrderedDescending;
            } else {
                return NSOrderedSame;
            }
        };
        
        // Sort the array in ascending order using a comparator block
        [list1 sortUsingComparator:ascendingComparison];
        [list2 sortUsingComparator:ascendingComparison];
        
        NSMutableDictionary *numberToOccurrences = [NSMutableDictionary dictionary];
        
        for (int i = 0; i < (int)[list1 count]; i++) {
            NSInteger numberOccurrences = 0;
            for (int j = 0; j < (int)[list2 count]; j++) {
                if (list1[i] == list2[j]) {
                    numberOccurrences += 1;
                }
            }
            
            [numberToOccurrences setObject:@(numberOccurrences) forKey:list1[i]];
        }
        
        NSInteger totalOccurrences = 0;
        NSEnumerator *enumerator = [numberToOccurrences keyEnumerator];
        id key;
        while ((key = [enumerator nextObject])) {
            NSNumber *value = numberToOccurrences[key];
            totalOccurrences += [key integerValue] * [value integerValue];
        }
        
        NSLog(@"%lu", totalOccurrences);
    }
    
    return 0;
}

File: day1/swift/day1-part2.swift
import Foundation

let directoryPath = FileManager.default.currentDirectoryPath

// Go back one directory (use ".." to navigate up)
let parentDirectoryPath = (directoryPath as NSString).deletingLastPathComponent  // Go one directory up
let filePath = parentDirectoryPath + "/input.txt"

let fileContents = try! String(contentsOfFile: filePath, encoding: .utf8)

let pairs = fileContents
    .components(separatedBy: "\n")
    .map { $0.split(separator: " ") }

var list1 = [Int]()
var list2 = [Int]()

for pair in pairs {
    list1.append(Int(pair[0])!)
    list2.append(Int(pair[1])!)
}

list1.sort()
list2.sort()

var similarityScore = 0

var numberToOccurrences: [Int: Int] = [:]
for num1 in list1 {
    var numberOccurrences = 0
    for num2 in list2 {
        if num2 == num1 {
            numberOccurrences += 1
        }
    }
    
    numberToOccurrences[num1] = numberOccurrences
}

for (key, value) in numberToOccurrences {
    similarityScore += key * value
}

print(similarityScore)
